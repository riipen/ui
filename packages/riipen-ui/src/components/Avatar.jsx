import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

class Avatar extends React.Component {
  static propTypes = {
    /**
     * Used to provide an alt attribute to the rendered img element.
     */
    alt: PropTypes.string,

    /**
     * Used to render icon or text elements inside the Avatar if src is not set.
     */
    children: PropTypes.node,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * The height and width size to render the Avatar at.
     */
    size: PropTypes.string.isRequired,

    /**
     * The src attribute for the img element.
     */
    src: PropTypes.string,

    /**
     * The shape of the avatar.
     */
    variant: PropTypes.oneOf(["circle", "rounded", "square"])
  };

  static defaultProps = {
    classes: [],
    size: "96px",
    variant: "circle"
  };

  getBorderSize() {
    const { size } = this.props;

    // Remove 'px' and get number of pixels
    const pixels = +size.substring(0, size.length - 2);

    switch (true) {
      case pixels < 70:
        return "2px";
      default:
        return "4px";
    }
  }

  static contextType = ThemeContext;

  render() {
    const { alt, children, classes, size, src, variant, ...other } = this.props;

    const theme = this.context;

    const className = clsx("avatar", variant, classes);

    return (
      <React.Fragment>
        {src ? (
          <img alt={alt} src={src} className={className} {...other} />
        ) : (
          <span className={className}>{children}</span>
        )}
        <style jsx>{`
          .avatar {
            align-items: center;
            background-color: ${theme.palette.grey[200]};
            border: ${this.getBorderSize()} solid ${theme.palette.common.white};
            display: inline-flex;
            flex-shrink: 0;
            font-family: ${theme.typography.fontFamily};
            font-size: calc(${size} * 0.4);
            height: ${size};
            justify-content: center;
            line-height: 1;
            overflow: hidden;
            user-select: none;
            width: ${size};
          }

          .circle {
            border-radius: 50%;
          }
          .rounded {
            border-radius: ${theme.shape.borderRadius.md};
          }
          .square {
            border-radius: 0px;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses()(Avatar);
