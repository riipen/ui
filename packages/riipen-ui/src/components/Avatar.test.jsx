import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import Avatar from "./Avatar";

describe("<Avatar>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Avatar />);
    } catch (e) {
      error = e;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Avatar />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new Avatar().type.defaultProps;

      const wrapper = mount(<Avatar />);

      expect(wrapper.find("Avatar").props().size).toEqual(defaultProps.size);
      expect(wrapper.find("Avatar").props().variant).toEqual(
        defaultProps.variant
      );
    });
  });

  describe("alt prop", () => {
    it("renders an img tag with alt value", () => {
      const altVariant = "Test String";
      const srcVariant = "/static/images/avatars/jordan-ell.png";

      const wrapper = mount(<Avatar alt={altVariant} src={srcVariant} />);

      expect(wrapper.find("img").props().alt).toEqual(altVariant);
    });
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const child = <div>ZZ</div>;

      const wrapper = mount(<Avatar>{child}</Avatar>);

      expect(
        wrapper
          .find(".inner")
          .children()
          .contains(child)
      ).toEqual(true);
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classVariant = ["newClass"];

      const wrapper = mount(<Avatar classes={classVariant} />);

      expect(
        wrapper
          .find("Avatar")
          .childAt(0)
          .hasClass(classVariant[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const wrapper = mount(<Avatar />);

      expect(wrapper.find("Avatar").props().classes).toEqual([
        "riipen",
        "riipen-avatar"
      ]);
    });
  });

  describe("size prop", () => {
    it("sets correct border size when size is above 70", () => {
      const sizeVariant = "75px";

      const wrapper = mount(<Avatar size={sizeVariant} />);

      expect(
        wrapper
          .find("Avatar")
          .instance()
          .getBorderSize()
      ).toEqual("4px");
    });

    it("sets correct border size when size is below 70", () => {
      const sizeVariant = "65px";

      const wrapper = mount(<Avatar size={sizeVariant} />);

      expect(
        wrapper
          .find("Avatar")
          .instance()
          .getBorderSize()
      ).toEqual("2px");
    });
  });

  describe("src prop", () => {
    it("renders an img tag with src value", () => {
      const srcVariant = "/static/images/avatars/jordan-ell.png";

      const wrapper = mount(<Avatar src={srcVariant} />);

      expect(wrapper.find("Avatar").find("img")).toHaveLength(1);
      expect(wrapper.find("img").props().src).toEqual(srcVariant);
    });
  });

  describe("variant prop", () => {
    it("sets variant class name with valid custom variant", () => {
      const variant = "rounded";

      const wrapper = mount(<Avatar variant={variant} />);

      expect(
        wrapper
          .find("Avatar")
          .childAt(0)
          .hasClass(variant)
      ).toEqual(true);
    });

    it("throws an error with with invalid variant", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<Avatar variant="star" />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });
});
