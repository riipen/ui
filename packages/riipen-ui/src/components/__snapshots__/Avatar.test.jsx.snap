// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Avatar> default props sets correct default props 1`] = `
<withClasses(Avatar)>
  <Avatar
    classes={
      Array [
        "riipen",
        "riipen-avatar",
      ]
    }
    size="96px"
    variant="circle"
  >
    <span
      className="jsx-3881544049 avatar circle riipen riipen-avatar"
    >
      <span
        className="jsx-3881544049 inner"
      />
    </span>
    <JSXStyle
      dynamic={
        Array [
          "#eeeeee",
          "4px",
          "#fff",
          "Roboto, Helvetica, Arial, sans-serif",
          "96px",
          "96px",
          "96px",
          "96px",
          "96px",
          "96px",
          "4px",
        ]
      }
      id="4017156812"
    />
  </Avatar>
</withClasses(Avatar)>
`;
